p2panda-rs/src/wasm.rs

```diff
@@ -0,0 +1,139 @@
+use std::convert::TryFrom;
+use std::panic;
+
+use console_error_panic_hook::hook as panic_hook;
+use serde::{Deserialize, Serialize};
+use wasm_bindgen::prelude::wasm_bindgen;
+use wasm_bindgen::JsValue;
+
+use crate::atomic::MessageFields as PandaMessageFields;
+use crate::atomic::{
+    Entry, EntrySigned, Hash, LogId, Message, MessageEncoded, MessageValue, SeqNum,
+};
+use crate::key_pair::KeyPair;
+
+macro_rules! jserr {
+    ($l:expr) => {
+        $l.map_err::<JsValue, _>(|err| js_sys::Error::new(&format!("{}", err)).into())?;
+    };
+}
+
+/// Sets a [`panic hook`] for better error messages in NodeJS or web browser.
+///
+/// [`panic hook`]: https://crates.io/crates/console_error_panic_hook
+#[wasm_bindgen(js_name = setWasmPanicHook)]
+pub fn set_wasm_panic_hook() {
+    panic::set_hook(Box::new(panic_hook));
+}
+
+#[wasm_bindgen]
+pub struct MessageFields(PandaMessageFields);
+
+#[wasm_bindgen]
+impl MessageFields {
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> Self {
+        Self(PandaMessageFields::new())
+    }
+
+    pub fn add(&mut self, name: String, value: JsValue) -> Result<(), JsValue> {
+        // @TODO: Add more types
+        let field = match value.as_string() {
+            Some(text) => Ok(MessageValue::Text(text)),
+            None => Err(js_sys::Error::new(&format!("Invalid value type"))),
+        }?;
+
+        jserr!(self.0.add(&name, field));
+
+        Ok(())
+    }
+}
+
+#[wasm_bindgen(js_name = encodeCreateMessage)]
+pub fn encode_create_message(schema: String, fields: MessageFields) -> Result<String, JsValue> {
+    let hash = jserr!(Hash::new(&schema));
+    let message = jserr!(Message::new_create(hash, fields.0));
+    let message_encoded = jserr!(MessageEncoded::try_from(&message));
+    Ok(message_encoded.as_str().to_owned())
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct SignEntryResult {
+    pub entry_encoded: String,
+    pub entry_hash: String,
+}
+
+#[wasm_bindgen(js_name = signEntry)]
+pub fn sign_entry(
+    key_pair: KeyPair,
```

Comment:
I think this is causing some issues on the JS end of things. After passing a KeyPair instance into signEntry() it no longer exists on the JS end. I guess this is a rust ownership thing. I tried:

`key_pair: &KeyPair,`

And that works, but I don't know if it is strange Rust....

Right now I am just passing in a copy of the keyPair which also works.

Reply:
> And that works, but I don't know if it is strange Rust....

Looks like a good solution! Passing over instances via reference is even better here. Do you want to make a commit?